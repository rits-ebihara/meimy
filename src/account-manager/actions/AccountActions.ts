import { Alert } from 'react-native';
import { getGenericPassword, setGenericPassword } from 'react-native-keychain';
import { Action, Dispatch } from 'redux';
import ShortId from 'shortid';

import { dateParser } from '../../eim-service/EIMServiceAdapter';
import { getConfig } from '../Config';
import { createInitAccountListState, IAccountListState } from '../states/IAccountLisState';
import { IAccountState } from '../states/IAccountState';
import { asyncLoadAccountListAfterShow } from './AccountListActions';
import { langProfile } from '../../LangProfile';

export const SET_ACCOUNT_ACTION = ShortId();

export interface ISetAccountAction extends Action {
    account: IAccountState;
}

export const createSetAccountAction = (account: IAccountState): ISetAccountAction => {
    return {
        account,
        type: SET_ACCOUNT_ACTION,
    };
};

const config = getConfig();

export const asyncSaveAccountAction
    = async (account: IAccountState, dispatch: Dispatch) => {
        try {
            const { siteName, siteDomain, authType, id, eimToken, lastConnect, userId, password } = account;
            // 拡張した他のプロパティが入らないように精査する
            const saveData: IAccountState = {
                authType,
                eimToken,
                id,
                lastConnect,
                password,
                siteDomain,
                siteName,
                userId,
            };
            // 現在のリストをロード
            const json = await getGenericPassword({ service: config.accountListServiceName });
            let accountList: IAccountListState = createInitAccountListState();
            if (!!json && typeof json !== 'boolean') {
                accountList = JSON.parse(json.password, dateParser) as IAccountListState;
            }
            // IDで検索しなければ追加する
            const existIndex = accountList.accounts.findIndex((a) => a.id === id);
            if (existIndex === -1) {
                accountList.accounts.push(saveData);
            } else {
                accountList.accounts[existIndex] = saveData;
            }
            await setGenericPassword('dummy', JSON.stringify(accountList), { service: config.accountListServiceName });
        } catch (error) {
            Alert.alert('error', langProfile.replaceLang('LK_MSG_saveError'));
            console.error(error);
        }
        // ステートを更新して、一覧に戻ったときに画面の表示が新しくなっているようにする
        await asyncLoadAccountListAfterShow(dispatch);
    };

export const REMOVE_ACCOUNT_ACTION = ShortId();

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface IRemoveAccountAction extends Action {
}

export const asyncRemoveAccountAction = async (
    targetId: string, dispatch: Dispatch, onSuccess: () => void) => {
    const save = async () => {
        try {
            // 現在のリストをロード
            const json = await getGenericPassword({ service: config.accountListServiceName });
            let accountList: IAccountListState = createInitAccountListState();
            if (!!json && typeof json !== 'boolean') {
                accountList = JSON.parse(json.password) as IAccountListState;
            } else {
                return;
            }
            accountList.accounts = accountList.accounts.filter((a) => a.id !== targetId);
            await setGenericPassword('dummy', JSON.stringify(accountList), { service: config.accountListServiceName });
        } catch (error) {
            Alert.alert('error', langProfile.replaceLang('LK_MSG_saveError'));
            console.error(error);
        }
    };
    await save();
    // ステートを更新して、一覧に戻ったときに画面の表示が新しくなっているようにする
    await asyncLoadAccountListAfterShow(dispatch);
    onSuccess();
};
