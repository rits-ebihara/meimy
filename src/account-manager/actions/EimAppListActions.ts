// import { Toast } from 'native-base';
import { NavigationScreenProp } from 'react-navigation';
import { Action, Dispatch } from 'redux';
import ShortId from 'shortid';

import { EIMServiceAdapter } from '../../eim-service/EIMServiceAdapter';
import { IEimApp } from '../states/IEimAppListState';
import INavigateController from './INavigateController';

interface ISourceEimApp {
    system: {};
    properties: {
        id: string;
        name: string;
        description: string;
    };
}

interface IGetAppListResponse {
    apps: ISourceEimApp[];
}

export const SET_APP_LIST = ShortId();
export interface ISetAppListAction extends Action {
    appList: IEimApp[];
}
export const createSetAppListAction = (appList: IEimApp[]): ISetAppListAction => {
    return {
        appList,
        type: SET_APP_LIST,
    };
};

export const LOAD_APP_LIST = ShortId();

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface ILoadAppListAction extends Action {
}
export const createLoadAppListAction =
    async (dispatch: Dispatch, navigateActions: INavigateController,
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        navigation: NavigationScreenProp<any>, onError: () => void): Promise<void> => {
        const { siteDomain, siteName, tokens, appKeyPrefix } = navigateActions.getLinkState();
        if (!siteDomain || !tokens || !appKeyPrefix) { return; }
        dispatch({
            type: LOAD_APP_LIST,
        });
        // アプリ一覧の取得
        await (async () => {
            const sa = new EIMServiceAdapter(siteDomain);
            try {
                const response = await sa.getAppList(tokens);
                const responseObj = JSON.parse(response.body) as IGetAppListResponse;
                const appList = responseObj.apps
                    .filter((a) => a.properties.id.startsWith(appKeyPrefix))
                    .map((sourceApp): IEimApp => {
                        const { properties } = sourceApp;
                        return {
                            appKey: properties.id,
                            appName: properties.name,
                            description: properties.description,
                            siteDomain,
                            siteName: siteName || siteDomain,
                            tokens,
                        };
                    });
                // リストに１件しかない場合
                if (appList.length === 1) {
                    const appKey = appList[0].appKey;
                    navigateActions.openApp({ appKey }, navigation);
                } else {
                    dispatch(createSetAppListAction(appList));
                }
            } catch (e) {
                console.error(e);
                onError();
            }
        })();
    };
