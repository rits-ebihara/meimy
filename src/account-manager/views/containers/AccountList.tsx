import { Body, Container, Content, Icon, List, ListItem, Right, Text } from 'native-base';
import React, { Component } from 'react';
import { NavigationScreenOptions } from 'react-navigation';
import { connect } from 'react-redux';

import { FloatingButton } from '../../../components/FloatingButton/FloatingButton';
import { ICombinedNavProps, IProps } from '../../../redux-helper/redux-helper';
import { createSetAccountAction } from '../../actions/AccountActions';
import { asyncLoadAccountListAfterShow } from '../../actions/AccountListActions';
import navigateController from '../../actions/NavigateActions';
import { getConfig } from '../../Config';
import { IAccountManagerState } from '../../IAccountManagerState';
import RoutePageNames from '../../RoutePageNames';
import { IAccountListState } from '../../states/IAccountLisState';
import { createInitAccountState, IAccountState } from '../../states/IAccountState';
import { IAuthState } from '../../states/IAuthStates';

const config = getConfig();
// eslint-disable-next-line @typescript-eslint/class-name-casing
export class _AccountList extends Component<ICombinedNavProps<IAccountListState>> {
    // 関数にしないと、config が反映されない状態で起動する
    public static navigationOptions = (): NavigationScreenOptions => ({
        headerStyle: {
            backgroundColor: config.colorPalets.$colorPrimary3,
        },
        headerTintColor: config.colorPalets.$invertColor,
        headerTitle: 'EIMサイト一覧',
    });
    public render = () => {
        // console.log(this.props.state.accounts);
        const { state } = this.props;
        const { theme } = config;
        const listItems = state.accounts.map((account) => {
            if (!account.id) { return null; }
            return (
                <ListItem noIndent key={account.id} button onPress={() => { this.onPressListItem(account); }}>
                    <Body>
                        <Text>{account.siteName}</Text>
                        <Text note>{account.siteDomain}</Text>
                    </Body>
                    <Right><Icon name="arrow-forward" /></Right>
                </ListItem>
            );
        });
        return (
            <Container>
                <Content>
                    <List>
                        {listItems}
                    </List>
                </Content>
                <FloatingButton
                    key="add_button"
                    primaryButtonColor={theme.btnPrimaryBg}
                    onPress={this.onPressAddButton} />
            </Container>
        );
    }
    public componentDidMount = async () => {
        await asyncLoadAccountListAfterShow(this.props.dispatch);
        await this.transferPage(this.props.state.authState);
    }
    private onPressListItem = (account: IAccountState) => {
        this.props.dispatch(createSetAccountAction(account));
        this.props.navigation.push(RoutePageNames.accountPageName);
    }
    private onPressAddButton = () => {
        this.props.dispatch(createSetAccountAction(createInitAccountState()));
        this.props.navigation.push(RoutePageNames.accountPageName);
    }

    private transferPage = async (authState: IAuthState) => {
        // ドメインが指定された場合、トークンがあればそれをセットする
        const account = this.props.state.accounts
            .find((a) => a.siteDomain === authState.siteDomain);
        if (!!account) {
            authState.tokens = account.eimToken;
            authState.siteName = account.siteName;
        }
        // this.props.navigation.navigate(accountListPageName);
        navigateController.clear();
        await navigateController.navigateForLink(
            this.props.state,
            authState,
            this.props.dispatch,
            this.props.navigation);
    }
}

const mapStateToProps = (state: IAccountManagerState): IProps<IAccountListState> => {
    return {
        state: state.accountList,
    };
};

export const __private__ = {
    mapStateToProps,
};

export default connect(mapStateToProps)(_AccountList);

// eslint-disable-next-line max-len
// adb shell am start -a android.intent.action.VIEW "eimmobile://accountmanager/?app=decision-flow\&domain=app-dev43.ope.azure.ricoh-eim.com"
// eslint-disable-next-line max-len
// adb shell am start -a android.intent.action.VIEW "eimapplink-accountmanager://accountmanager/?mapp=decision-flow\&domain=app-dev54.ope.azure.ricoh-eim.com\&appprefix=DWS\&link=https%3A%2F%2Frits3.eim.ricoh.com%2F%23%2Fapps%2FdocManagementApp%2Fdocuments%2F9dd4435823554142a069ce9ee81b1e12"
