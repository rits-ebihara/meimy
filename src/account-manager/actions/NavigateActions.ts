import Clone from 'clone';
import { Linking } from 'react-native';
import { NavigationScreenProp } from 'react-navigation';
import { Dispatch } from 'redux';
import UrlParse from 'url-parse';

import { EIMServiceAdapter } from '../../eim-service/EIMServiceAdapter';
import { getConfig } from '../Config';
import { getEimAccount } from '../EimAccount';
import { IAuthAppQuery } from '../IEimAccount';
import RoutePageNames from '../RoutePageNames';
import { IAccountListState } from '../states/IAccountLisState';
import { createInitAccountState } from '../states/IAccountState';
import { IAuthState } from '../states/IAuthStates';
import { createSetAccountAction } from './AccountActions';
import { createSetAuthState } from './AccountListActions';
import { createSetAppListAction } from './EimAppListActions';
import INavigateController from './INavigateController';

export class NavigateController implements INavigateController {
    public parentMainPage?: string;
    public parentNavParams?: { [key: string]: any };
    private linkStates: IAuthState = {};
    private transferAccountPage = (
        accountListState: IAccountListState,
        dispatch: Dispatch,
        navigation: NavigationScreenProp<any>,
        replace: boolean,
        siteDomain: string,
    ) => {
        const accountList = accountListState.accounts;
        let account = accountList.find((a) => a.siteDomain === this.linkStates.siteDomain);
        if (!account) {
            // なければアカウント新規作成画面に遷移する
            account = createInitAccountState();
            const name = siteDomain;
            account.siteName = name;
            account.siteDomain = name;
        }
        dispatch(createSetAccountAction(account));
        if (replace) {
            navigation.replace(RoutePageNames.accountPageName);
        } else {
            navigation.navigate(RoutePageNames.accountPageName);
        }
    };
    private determinedAppAndDomain = async (
        siteDomain: string,
        navigation: NavigationScreenProp<any>,
        accountListState: IAccountListState,
        dispatch: Dispatch,
        replace: boolean,
    ): Promise<boolean> => {
        // トークンがある
        const sa = new EIMServiceAdapter(siteDomain);
        const connected = await sa.validateToken(this.linkStates.tokens || []);
        // トークン検証
        if (!!connected) {
            // 成功
            // 呼び出しアプリを起動
            await this.openApp({}, navigation);
            return true;
        } else {
            // 失敗
            // アカウント画面に遷移する
            this.transferAccountPage(
                accountListState, dispatch, navigation, replace, siteDomain);
            return true;
        }
    }
    private deteminedAppOnly = async (
        domain: string,
        replace: boolean,
        navigation: NavigationScreenProp<any>,
        dispatch: Dispatch,
        accountListState: IAccountListState,
    ) => {
        const sa = new EIMServiceAdapter(domain);
        // 接続済みの場合
        // トークンがある場合は、検証する
        const connected = await sa.validateToken(this.linkStates.tokens || []);
        // 検証の結果
        if (connected) {
            // 認証が成功の場合
            // アプリ一覧を取得
            // アプリ一覧の画面に遷移する
            const moveTo = (replace) ? navigation.replace : navigation.navigate;
            if (this.linkStates.appKeyPrefix) {
                dispatch(createSetAppListAction([]));
                moveTo(RoutePageNames.appListPageName);
            } else {
                moveTo(RoutePageNames.accountListPageName);
            }
            return true;
        } else {
            // 失敗の場合、アカウントの画面に遷移する
            this.transferAccountPage(
                accountListState, dispatch, navigation, replace, domain);
        }
        return true;
    }
    public navigateForLink = async (
        accountListState: IAccountListState,
        pLinkState: IAuthState,
        dispatch: Dispatch,
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        navigation: NavigationScreenProp<any>,
        replace: boolean = false
    ): Promise<boolean> => {
        this.linkStates = Object.assign({}, this.linkStates, pLinkState);
        if (!!this.linkStates.siteDomain && !!this.linkStates.appKey) {
            // ドメインとアプリが決まっている
            return await this.determinedAppAndDomain(
                this.linkStates.siteDomain,
                navigation,
                accountListState,
                dispatch,
                replace);
        } else if (!!this.linkStates.siteDomain) {
            // サイトドメインが決まっている
            // トークンがある
            return await this.deteminedAppOnly(
                this.linkStates.siteDomain,
                replace,
                navigation,
                dispatch,
                accountListState,
            );
        } else {
            // ドメインの指定がない場合は、なにもしない
            return false;
        }
    }
    public clear = () => {
        this.linkStates = {};
    }
    public getLinkState = () => {
        return Clone(this.linkStates);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    public openApp = async (pLinkState: IAuthState, navigation: NavigationScreenProp<any>) => {
        this.linkStates = Object.assign({}, this.linkStates, pLinkState);
        const eimAccount = getEimAccount();
        eimAccount.appKey = this.linkStates.appKey || '';
        eimAccount.domain = this.linkStates.siteDomain || '';
        eimAccount.siteName = this.linkStates.siteName || '';
        eimAccount.eimTokens = this.linkStates.tokens || [];
        await eimAccount.save();
        await eimAccount.loadUser();
        if (this.parentMainPage) {
            navigation.navigate(this.parentMainPage, this.parentNavParams);
        }
    }
    public openAccountManager = async (
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        navigation: NavigationScreenProp<any>,
        dispatch: Dispatch,
        link?: string,
        hash?: string
    ) => {
        let appKey: string | undefined;
        let domain: string | undefined;
        // 部分適用
        const open = (appKey?: string, domain?: string) => {
            return this._openAccountManager(
                dispatch, navigation, link, hash,
                appKey, domain);
        }
        // メールなど文書を開くリンクから開く
        if (!!link && !!hash) {
            const linkUrl = UrlParse(link);
            domain = linkUrl.host;
            const paths = hash.split('/');
            appKey = paths.length === 5 ? paths[2] : undefined;
            open(appKey, domain);
            return;
        }
        const eimAccount = getEimAccount();
        // 前回アクセスしたサイトを取得する
        await eimAccount.load().then((lastAccount) => {
            // 認証アプリを起動する
            if (!!lastAccount) {
                open(lastAccount.appKey, lastAccount.domain);
            } else {
                open();
            }
        });
    }
    private _openAccountManager = async (
        dispatch: Dispatch, navigation: NavigationScreenProp<any>,
        link?: string, hash?: string,
        appKey?: string, domain?: string
    ) => {
        const config = getConfig();

        const authState: IAuthState = {
            appKeyPrefix: config.appKeyPrefix,
        };
        const accountManagerUrl = UrlParse('eimapplink-accountmanager://accountmanager/');
        const query: IAuthAppQuery = {
            appprefix: config.appKeyPrefix,
            mapp: config.appKeyPrefix,
        };
        if (!!appKey && !!domain) {
            query.appkey = appKey;
            query.domain = domain;
            authState.appKey = appKey;
            authState.siteDomain = domain;
        }
        if (!!link) {
            const linkUrl = UrlParse(link);
            linkUrl.set('hash', hash);
            query.link = linkUrl.href;
            authState.link = linkUrl.href;
        }

        accountManagerUrl.set('query', query);
        await Linking.canOpenURL(accountManagerUrl.href).then((result) => {
            if (result) {
                Linking.openURL(accountManagerUrl.href);
            } else {
                this.parentMainPage = config.startPage;
                if (!!hash) {
                    const paths = hash.split('/');
                    this.parentNavParams = {
                        parameter: paths.length === 5 ? paths[4] : undefined,
                    };
                }
                dispatch(createSetAuthState(authState));
                navigation.navigate(RoutePageNames.authPageName);
            }
        });
    }
}
export default new NavigateController();
