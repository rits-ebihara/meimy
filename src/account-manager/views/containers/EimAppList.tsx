import { Body, Container, ListItem, Spinner, Text, Toast } from 'native-base';
import React, { Component } from 'react';
import { BackHandler } from 'react-native';
import { connect } from 'react-redux';

import { ICombinedNavProps, IProps } from '../../../redux-helper/redux-helper';
import { createLoadAppListAction, createSetAppListAction } from '../../actions/EimAppListActions';
import NavigateActions from '../../actions/NavigateActions';
import { getConfig } from '../../Config';
import { IAccountManagerState } from '../../IAccountManagerState';
import IEimAppList from '../../states/IEimAppListState';

const config = getConfig();

class EimAppList extends Component<ICombinedNavProps<IEimAppList>> {
    public static navigationOptions = () => {
        return {
            headerStyle: {
                backgroundColor: config.colorPalets.$colorPrimary3,
            },
            headerTintColor: config.colorPalets.$invertColor,
            headerTitle: 'アプリ一覧',
        };
    }
    public constructor(props: ICombinedNavProps<IEimAppList>) {
        super(props);
        BackHandler.addEventListener('hardwareBackPress', this.onPressBackButton);
    }
    public render = () => {
        const appListBox = this.props.state.appList.map((app) => {
            return (
                <ListItem noIndent key={app.appKey} onPress={this.onPressItem.bind(this, app.appKey)}>
                    <Body>
                        <Text>{app.appName}</Text>
                        <Text note>{app.description}</Text>
                    </Body>
                </ListItem>
            );
        });
        return (
            <Container>
                {this.props.state.loading ? <Spinner color="blue" /> : appListBox}
            </Container>
        );
    }
    public componentDidMount = () => {
        createLoadAppListAction(this.props.dispatch, NavigateActions, this.props.navigation, this.onNetworkError);
    }
    public onPressItem = (appKey: string) => {
        NavigateActions.openApp({ appKey }, this.props.navigation);
    }

    private onNetworkError = () => {
        Toast.show({
            text: 'ネットワークエラーが発生しました。',
            type: 'warning',
        });
        this.props.dispatch(createSetAppListAction([]));
    }

    private onPressBackButton = () => {
        this.props.navigation.pop();
        return true;
    }
}

const mapStateToProps = (state: IAccountManagerState): IProps<IEimAppList> => {
    return {
        state: state.appList,
    };
};

export const __private__ = {
    EimAppList: EimAppList,
    mapStateToProps,
};

export default connect(mapStateToProps)(EimAppList);
