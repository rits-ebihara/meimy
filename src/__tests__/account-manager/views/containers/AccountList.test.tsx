import Enzyme from 'enzyme';
import EnzymeAdapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import React from 'react';
import { mocked } from 'ts-jest/utils';

import { SET_ACCOUNT_ACTION } from '../../../../account-manager/actions/AccountActions';
import { asyncLoadAccountListAfterShow } from '../../../../account-manager/actions/AccountListActions';
import navigateController from '../../../../account-manager/actions/NavigateActions';
import { IAccountListState } from '../../../../account-manager/states/IAccountLisState';
import { IAccountState } from '../../../../account-manager/states/IAccountState';
import { IAuthState } from '../../../../account-manager/states/IAuthStates';
import { __private__, _AccountList } from '../../../../account-manager/views/containers/AccountList';
import { ICombinedNavProps } from '../../../../redux-helper/redux-helper';

Enzyme.configure({ adapter: new EnzymeAdapter() });

jest.mock('../../../../account-manager/actions/AccountListActions.ts');

jest.mock('../../../../account-manager/actions/NavigateActions.ts');

const getProp = (): ICombinedNavProps<IAccountListState> => {
    return {
        navigation: {
            setParams: jest.fn(),
            pop: jest.fn(),
            navigate: jest.fn(),
            push: jest.fn(),
        } as any,
        dispatch: jest.fn(),
        state: {
            accounts: [],
            authState: {
                appKey: 'app-key',
                appKeyPrefix: 'app-prefix',
                mobileAppKey: 'app-m-key',
                siteDomain: 'site-domain',
                siteName: 'site-name',
                tokens: ['token'],
            },
        },
    };
};

const getAccountList = (): IAccountState[] => {
    return [
        {
            authType: 'password',
            eimToken: [],
            id: 'a001',
            lastConnect: null,
            siteDomain: 'site-domain',
            siteName: 'site-name',
        },
        {
            authType: 'password',
            eimToken: [],
            id: 'a002',
            lastConnect: null,
            siteDomain: 'site-domain2',
            siteName: 'site-name2',
        },
    ];
};

describe('init', () => {
    beforeEach(() => {
        mocked(asyncLoadAccountListAfterShow).mockClear();
    })
    test('blank list', async () => {
        const props = getProp();
        const wrapper = Enzyme.shallow(<_AccountList {...props} />);
        expect(toJson(wrapper)).toMatchSnapshot();
        expect(asyncLoadAccountListAfterShow).toBeCalled();
        // expect(_AccountList.prototype['transferPage']).toBeCalled();
    });
    test('exist list', async () => {
        const props = getProp();
        props.state.accounts = getAccountList();
        props.state.accounts.push({
            authType: 'password',
            eimToken: [],
            id: null,
            lastConnect: null,
            siteDomain: 'site-domain2',
            siteName: 'site-name2',
        });
        const wrapper = Enzyme.shallow(<_AccountList {...props} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});

describe('events', () => {
    beforeEach(() => {
    });
    test('onPressListItem', () => {
        const props = getProp();
        props.state.accounts = getAccountList();
        const wrapper = Enzyme.shallow(<_AccountList {...props} />);
        const listItem = wrapper.findWhere(a => a.key() === 'a001');
        listItem.simulate('press');
        expect(props.dispatch).toBeCalledWith({
            account: props.state.accounts[0],
            type: SET_ACCOUNT_ACTION,
        });
        expect(props.navigation.push).toBeCalledWith('accountPage');
    });
    test('onPressListItem', () => {
        const props = getProp();
        props.state.accounts = getAccountList();
        const wrapper = Enzyme.shallow(<_AccountList {...props} />);
        const addButton = wrapper.findWhere(a => a.key() === 'add_button');
        addButton.simulate('press');
        expect(props.dispatch).toBeCalledWith({
            account: {
                authType: 'o365',
                eimToken: [],
                id: null,
                lastConnect: null,
                siteDomain: '',
                siteName: '',
            },
            type: SET_ACCOUNT_ACTION,
        });
        expect(props.navigation.push).toBeCalledWith('accountPage');
    });
});

describe('transferPage', () => {
    beforeEach(() => {
        mocked(navigateController.navigateForLink).mockClear();
    })
    test('', async () => {
        const props = getProp();
        props.state.accounts = getAccountList();
        const wrapper = Enzyme.shallow(<_AccountList {...props} />);
        const instance = wrapper.instance() as _AccountList;
        const authState: IAuthState = {
            appKey: 'app-key',
            appKeyPrefix: 'app-prefix',
            mobileAppKey: 'app-m-key',
            siteDomain: 'site-domain',
            siteName: 'site-name',
            tokens: ['token'],
        };
        await instance['transferPage'](authState);
        expect(navigateController.navigateForLink).toBeCalled();
    });
});

describe('mapStateToProps', () => {
    test('', () => {
        const state = {
            accountList: getAccountList(),
            prop1: 'a',
            prop2: 'b',
        };
        const result = __private__.mapStateToProps(state as any);
        expect(result).toStrictEqual({
            state: state.accountList,
        });
    });
});

describe('navigationOptions', () => {
    test('', () => {
        const options = _AccountList.navigationOptions();
        expect(options).toEqual({
            headerStyle: {
                backgroundColor: '#162756',
            },
            headerTintColor: '#fff',
            headerTitle: 'EIMサイト一覧',
        });
    })
});
