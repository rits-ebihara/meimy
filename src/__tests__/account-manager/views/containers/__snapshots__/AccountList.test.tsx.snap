// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`init blank list 1`] = `
<Styled(Container)>
  <Styled(Content)>
    <Styled(List) />
  </Styled(Content)>
  <FloatingButton
    disable={false}
    iconColor="white"
    iconName="add"
    iconType="Ionicons"
    key="add_button"
    label=""
    marginHorizontal={32}
    marginVertical={32}
    onPress={[Function]}
    position="bottom-right"
    primaryButtonColor="#000099"
    primaryButtonSize={56}
    secondaryButtonSize={48}
  />
</Styled(Container)>
`;

exports[`init exist list 1`] = `
<Styled(Container)>
  <Styled(Content)>
    <Styled(List)>
      <Styled(ListItem)
        button={true}
        key="a001"
        noIndent={true}
        onPress={[Function]}
      >
        <Styled(Body)>
          <Styled(Text)>
            site-name
          </Styled(Text)>
          <Styled(Text)
            note={true}
          >
            site-domain
          </Styled(Text)>
        </Styled(Body)>
        <Styled(Right)>
          <Styled(Icon)
            name="arrow-forward"
          />
        </Styled(Right)>
      </Styled(ListItem)>
      <Styled(ListItem)
        button={true}
        key="a002"
        noIndent={true}
        onPress={[Function]}
      >
        <Styled(Body)>
          <Styled(Text)>
            site-name2
          </Styled(Text)>
          <Styled(Text)
            note={true}
          >
            site-domain2
          </Styled(Text)>
        </Styled(Body)>
        <Styled(Right)>
          <Styled(Icon)
            name="arrow-forward"
          />
        </Styled(Right)>
      </Styled(ListItem)>
    </Styled(List)>
  </Styled(Content)>
  <FloatingButton
    disable={false}
    iconColor="white"
    iconName="add"
    iconType="Ionicons"
    key="add_button"
    label=""
    marginHorizontal={32}
    marginVertical={32}
    onPress={[Function]}
    position="bottom-right"
    primaryButtonColor="#000099"
    primaryButtonSize={56}
    secondaryButtonSize={48}
  />
</Styled(Container)>
`;
